Fix errors while orientation covariance visualization in CovarianceVi…

UniBwTAS opened the pull request:
Hello,
this PR fixes some problems when visualizing the orientation uncertainty/covariance.
Steps to reproduce:

Execute in bash/zsh: `rostopic pub "/test" geometry_msgs/PoseWithCovarianceStamped '{header: {stamp: 0, frame_id: "map"}, pose: {pose: {position: {x: 0, y: 0, z: 0}, orientation: {x: 0, y: 0, z: 0, w: 1}}, covariance: [1, 0, 0, 0, 0, 0, 0, 0.1, 0, 0, 0, 0, 0, 0, 0.1, 0, 0, 0, 0, 0, 0, 0.01, 0, 0, 0, 0, 0, 0, 0.2, 0, 0, 0, 0, 0, 0, 0.01]}}'`

This publishes a pose with following orientation covarariance (roll, pitch, yaw):
```
0.01 0 0
0 0.2 0
0 0 0.01
```
This leads to the following visualization in RVIZ:
![Screenshot from 2020-08-24 17-08-48](https://user-images.githubusercontent.com/11663218/91068196-33cd4580-e634-11ea-80af-3fa27579e2d2.png)

The problem is that the covariance ellipse of the z-axis (blue) has the wrong orientation. I think it should be rotated by 90 degrees. The variance is large only for the y-axis (green) while the other two are very small. When one imagines it rotates around the green y-axis then the blue z-axis goes back and forth and not left and right.

With my PR it looks like this:
![Screenshot from 2020-08-24 17-27-14](https://user-images.githubusercontent.com/11663218/91069281-7fccba00-e635-11ea-84c5-8fe7357cfd56.png)

Another problem is the following:

Execute in bash/zsh: `rostopic pub "/test" geometry_msgs/PoseWithCovarianceStamped '{header: {stamp: 0, frame_id: "map"}, pose: {pose: {position: {x: 0, y: 0, z: 0}, orientation: {x: 0, y: 0, z: 0, w: 1}}, covariance: [1, 0, 0, 0, 0, 0, 0, 0.1, 0, 0, 0, 0, 0, 0, 0.1, 0, 0, 0, 0, 0, 0, 0.01, 0, 0, 0, 0, 0, 0, 0.2, 0.1, 0, 0, 0, 0, 0.1, 0.2]}}'`

This publishes a pose with following orientation covarariance (roll, pitch, yaw):
```
0.01 0 0
0 0.2 0.1
0 0.1 0.2
```
This leads to the following visualization in RVIZ:
![Screenshot from 2020-08-24 17-20-57](https://user-images.githubusercontent.com/11663218/91068438-94f51900-e634-11ea-9f58-b3ed997f92f2.png)

Again the the covariance ellipse of the blue z-axis should be rotated by 90 degrees. But additionally we now have high variances of y- **and** z-axis as well as non-diagonal entries in the covariance matrix, which means that y and z rotations are highly correlated. Assuming the angles are positive in counter-clockwise direction then the ellipse of the x-axis should go from top-left (positive angles) to bottom-right (negative angles).

With my PR it looks like this:
![Screenshot from 2020-08-24 17-29-32](https://user-images.githubusercontent.com/11663218/91070769-865c3100-e637-11ea-9173-d90603df8f8d.png)

UniBwTAS commented:
Further examples:
`rostopic pub "/test" geometry_msgs/PoseWithCovarianceStamped '{header: {stamp: 0, frame_id: "map"}, pose: {pose: {position: {x: 0, y: 0, z: 0}, orientation: {x: 0, y: 0, z: 0, w: 1}}, covariance: [1, 0, 0, 0, 0, 0, 0, 0.1, 0, 0, 0, 0, 0, 0, 0.1, 0, 0, 0, 0, 0, 0, 0.2, 0, 0.1, 0, 0, 0, 0, 0.01, 0, 0, 0, 0, 0.1, 0, 0.2]}}'`

This publishes a pose with following orientation covarariance (roll, pitch, yaw):

```
0.2 0 0.1
0 0.01 0
0.1 0 0.2
```

After PR looks like:
![Screenshot from 2020-08-24 19-49-27](https://user-images.githubusercontent.com/11663218/91078670-43a05600-e643-11ea-9552-9ef66b6e37e8.png)

`rostopic pub "/test" geometry_msgs/PoseWithCovarianceStamped '{header: {stamp: 0, frame_id: "map"}, pose: {pose: {position: {x: 0, y: 0, z: 0}, orientation: {x: 0, y: 0, z: 0, w: 1}}, covariance: [1, 0, 0, 0, 0, 0, 0, 0.1, 0, 0, 0, 0, 0, 0, 0.1, 0, 0, 0, 0, 0, 0, 0.2, 0.1, 0, 0, 0, 0, 0.1, 0.2, 0, 0, 0, 0, 0, 0, 0.01]}}'`


This publishes a pose with following orientation covarariance (roll, pitch, yaw):

```
0.2 0.1 0
0.1 0.2 0
0 0 0.01
```

After PR looks like:
![Screenshot from 2020-08-24 19-53-08](https://user-images.githubusercontent.com/11663218/91079078-dccf6c80-e643-11ea-87ee-45e1580a7ae9.png)

I think everything looks good now.

UniBwTAS commented:
Maybe related to: #1191

rhaschke commented:
@Ellon, could you please review this, as you initially implemented covariance visualization?
I'm not sure what's the intended visualization. But I agree with @UniBwTAS that it seems to be inconsistent with the present orientation.

Ellon commented:
I've completely forgot to address this. I'm sorry about that. I'll answer
as soon as I come back from holidays.

On Tue, Aug 25, 2020, 22:31 Robert Haschke <notifications@github.com> wrote:

> @Ellon <https://github.com/Ellon>, could you please review this, as you
> initially implemented covariance visualization?
> I'm not sure what's the intended visualization. But I agree with @UniBwTAS
> <https://github.com/UniBwTAS> that it seems to be inconsistent with the
> present orientation.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/ros-visualization/rviz/pull/1540#issuecomment-680254261>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAQ2VD3CAG3GA5R2P5DIJY3SCQNQXANCNFSM4QJV5I5Q>
> .
>

rhaschke commented:
@Ellon: Ping.

Ellon commented:
Hi @UniBwTAS and @rhaschke, sorry for the late reply.

@UniBwTAS I debugged all your examples "by hand" and it seems there's indeed a bug in the current code, at least for the cylinder on the z-axis. Could you show images "before PR" for your further examples? I'm having problems to compile and test noetic-devel on my current system and I cannot see how they were behaving before the PR. I imagine this would explain why you are proposing a change to `orientation_offset_node_[kRoll]` as well ?

While I'm at it, I'll investigate the relationship with #1191 .

UniBwTAS commented:
Hello @Ellon 

I updated the above "further examples" post. As far as I remember for actual visualization the change at the X Axis makes no difference. I just changed it to make it consistent such that the surface normal of the top points away from the center, I think. However this is not visual as the top and bottom look the same. I can remove the change for the X axis.

rhaschke commented:
@Ellon, @UniBwTAS: What's the status of this PR?

rhaschke commented:
@Ellon, @UniBwTAS, and @ViktorWalter: Could please review this PR (#1540) and #1576? If you all agree with the proposed solution(s), I'will be happy to merge them. Thanks a lot in advance.

ViktorWalter commented:
I have looked at #1540 - I can replicate the issue described here - the visualization of the "shape" attached to "Z" axis is statically rotated by 90 degrees and this commit fixes it. However, the change in line 276 as far as I can tell does nothing, as the "Covariance ellipses" are symmetrical. Still, I suggest merging #1540. This should be fully compatible with also merging #1576.

rhaschke commented:
Thanks, @ViktorWalter for your feedback. I will await the feedback.

Ellon commented:
> @Ellon, @UniBwTAS: What's the status of this PR?

Hi @rhaschke ! My apologies for being silent about this PR.

By looking at the commit and the examples @UniBwTAS gave it seems to be correcting the problem I didn't see at the time, but today it's not straight forward for me to test this PR in my current system (I'm using an hybrid of old ROS indigo with some other in-house things) so I couldn't run some tests with this PR by myself. I'll give myself these weeks evenings to have noetic running in a docker and test this PR.

rhaschke commented:
> I'll give myself this week evenings to have noetic running in a docker and test this PR.

Thanks @Ellon for helping to review this. Please also consider #15476.

Ellon commented:
> Please also consider #1576.

OK

Small update: I managed to have rviz from noetic-devel compiled inside a docker from ubuntu focal, but I'm unable to run rviz due to something related with GLX/nvidia/X11/cuda. I'll continue trying tomorrow, but if anyone knows what could be the problem please let me know.

rhaschke commented:
Running rviz from a docker container requires some extra work as described in section "Running MoveIt container" of https://moveit.ros.org/install/docker. We do have a script `gui-docker` to facilitate the task.

Ellon commented:
> We do have a script gui-docker to facilitate the task.

Thanks for the info. I'll take a look at it tonight.

rhaschke commented:
@Ellon, did you manage to review this PR and #1576?
